compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -323,17 +323,50 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
             }
         }
+        //issue 108713 forgotten writer warning
+        let is_write_macro = sugg_span
+            .ctxt()
+            .outer_expn_data()
+            .macro_def_id
+            .map_or(false, |def_id| tcx.is_diagnostic_item(sym::write_macro, def_id));
+        let is_write_fmt = item_name.name == Symbol::intern("write_fmt");
+        let mut err;
+        if is_write_fmt && is_write_macro && args.is_some() {
+            let arg_span = args.unwrap().0.span;
+            let writer_note;
+            match args.unwrap().0.kind {
+                ExprKind::Lit(_) => {
+                    writer_note = Some("A writer is needed before this format string")
+                }
+                _ => writer_note = None,
+            };
 
-        let mut err = struct_span_err!(
-            tcx.sess,
-            span,
-            E0599,
-            "no {} named `{}` found for {} `{}` in the current scope",
-            item_kind,
-            item_name,
-            rcvr_ty.prefix_string(self.tcx),
-            ty_str_reported,
-        );
+            err = struct_span_err!(
+                tcx.sess,
+                args.unwrap().0.span,
+                E0599,
+                "cannot write into '{}'",
+                ty_str_reported,
+            );
+            err.span_note(
+                arg_span,
+                "must implement 'io::Write', 'fmt::Write', or have a write_fmt method",
+            );
+            if writer_note.is_some() {
+                err.span_help(arg_span.shrink_to_lo(), writer_note.unwrap());
+            }
+        } else {
+            err = struct_span_err!(
+                tcx.sess,
+                span,
+                E0599,
+                "no {} named `{}` found for {} `{}` in the current scope",
+                item_kind,
+                item_name,
+                rcvr_ty.prefix_string(self.tcx),
+                ty_str_reported,
+            );
+        }
```

Comment:
Maybe this would be a good opportunity to migrate this into a [diagnostic struct](https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-structs.html)...

Reply:
Sorry not sure I'm understanding correctly, are you suggesting moving the missing writer portion into a diagnostic struct or moving the entire report_no_match_method_error into a diagnostic struct?

Reply:
Probably both, into different diagnostic structs

Reply:
Gotcha, thanks

Reply:
If the other changes I made look fine I can get started working on this